File: inspect_translations.py
#!/usr/bin/env python3
import os
import json
import re
from pathlib import Path
from collections import defaultdict

LOCALES_DIR = Path(__file__).parent / 'public' / 'locales'
SRC_DIR = Path(__file__).parent / 'src'
LANGUAGES = ['en', 'fr', 'de', 'it']
NS_MAPPING = {
    'home': 'pages/home',
    'blog': 'pages/blog',
    'about': 'pages/about',
    'faq': 'pages/faq',
    'facilities': 'pages/facilities',
    'professionals': 'pages/professionals',
    'blogArticles': 'pages/blog',
    'contact': 'pages/contact',
    'privacy': 'legal/privacy',
    'terms': 'legal/terms',
    'common': 'common',
    'admin': 'dashboard/admin',
    'dashboard': 'dashboard/dashboard',
    'calendar': 'dashboard/calendar',
    'dashboardPersonal': 'dashboard/personalDashboard',
    'dashboardProfile': 'dashboard/profile',
    'messages': 'dashboard/messages',
    'marketplace': 'dashboard/marketplace',
    'organization': 'dashboard/organization',
    'payroll': 'dashboard/payroll',
    'contracts': 'dashboard/contracts',
    'dropdowns': 'dropdowns',
    'validation': 'validation',
    'auth': 'auth',
    'pages/faq': 'pages/faq',
    'sitemap': 'pages/sitemap'
}

def get_all_files(directory):
    files = []
    for root, dirs, filenames in os.walk(directory):
        if 'node_modules' in root or '.git' in root:
            continue
        for filename in filenames:
            if filename.endswith(('.js', '.jsx')):
                files.append(os.path.join(root, filename))
    return files

def extract_translation_keys(content):
    keys = set()
    
    patterns = [
        r"t\(['\"`]([^'\"`]+)['\"`]",
        r"t\(['\"`]([^'\"`]+)['\"`]\s*,\s*['\"`]",
        r"i18n\.t\(['\"`]([^'\"`]+)['\"`]",
        r"i18next\.t\(['\"`]([^'\"`]+)['\"`]",
        r"\.t\(['\"`]([^'\"`]+)['\"`]",
    ]
    
    for pattern in patterns:
        matches = re.finditer(pattern, content)
        for match in matches:
            key = match.group(1)
            if key and '{{' not in key and '${' not in key and not key.startswith('http') and len(key) > 0:
                keys.add(key)
    
    return keys

def load_locale_file(lang, namespace):
    file_path = LOCALES_DIR / lang / f"{namespace}.json"
    if file_path.exists():
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                return json.load(f)
        except Exception as e:
            print(f"Warning: Error loading {file_path}: {e}")
            return None
    return None

def get_all_keys_from_object(obj, prefix=''):
    keys = []
    if isinstance(obj, dict):
        for key, value in obj.items():
            full_key = f"{prefix}.{key}" if prefix else key
            if isinstance(value, dict) and not isinstance(value, list):
                keys.extend(get_all_keys_from_object(value, full_key))
            else:
                keys.append(full_key)
    return keys

def get_nested_value(obj, key_path):
    keys = key_path.split('.')
    value = obj
    for key in keys:
        if isinstance(value, dict) and key in value:
            value = value[key]
        else:
            return None
    return value

def parse_translation_key(key):
    if ':' in key:
        colon_index = key.index(':')
        namespace = key[:colon_index].strip()
        key_path = key[colon_index + 1:].strip()
        return {'namespace': namespace, 'key': key_path}
    return {'namespace': 'common', 'key': key}

def find_hardcoded_strings(content):
    hardcoded = []
    lines = content.split('\n')
    
    suspicious_keywords = [
        'title', 'label', 'placeholder', 'text', 'message', 'error', 'success',
        'warning', 'info', 'description', 'button', 'submit', 'cancel', 'save',
        'delete', 'edit', 'add', 'remove', 'close', 'open', 'confirm', 'reject',
        'approve', 'decline', 'accept', 'logout', 'login', 'register', 'forgot',
        'reset', 'change', 'profile', 'settings', 'help', 'support', 'contact',
        'about', 'terms', 'privacy', 'language', 'theme', 'notifications', 'messages',
        'dashboard', 'home', 'back', 'next', 'continue', 'finish', 'required',
        'optional', 'select', 'search', 'filter', 'sort', 'show', 'hide', 'view',
        'preview', 'download', 'upload', 'expand', 'collapse', 'enable', 'disable',
        'activate', 'deactivate'
    ]
    
    for line_num, line in enumerate(lines, 1):
        if line.strip().startswith('//') or line.strip().startswith('*'):
            continue
        
        for keyword in suspicious_keywords:
            pattern = rf"{keyword}\s*[:=]\s*['\"`]([A-Z][^'\"`]{{3,}})['\"`]"
            matches = re.finditer(pattern, line, re.IGNORECASE)
            for match in matches:
                text = match.group(1)
                if (text and 
                    'http' not in text and 
                    'className' not in text and 
                    'id=' not in text and
                    'src=' not in text and
                    len(text) > 3 and
                    '{{' not in text and
                    '${' not in text):
                    hardcoded.append({
                        'line': line_num,
                        'text': text[:100],
                        'context': line.strip()[:150]
                    })
    
    return hardcoded

def main():
    print('🔍 Starting translation inspection...\n')
    
    all_locale_keys = {}
    used_keys = set()
    missing_keys = defaultdict(lambda: defaultdict(list))
    legacy_keys = defaultdict(lambda: defaultdict(list))
    hardcoded_strings = {}
    
    print('📚 Loading locale files...')
    for lang in LANGUAGES:
        all_locale_keys[lang] = {}
        for ns, file_path in NS_MAPPING.items():
            data = load_locale_file(lang, file_path)
            if data:
                all_locale_keys[lang][ns] = data
        
        common_file = load_locale_file(lang, 'common')
        if common_file:
            all_locale_keys[lang]['common'] = common_file
    
    print('🔎 Scanning source files for translation keys...')
    source_files = get_all_files(SRC_DIR)
    print(f'Found {len(source_files)} source files\n')
    
    for file_path in source_files:
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                content = f.read()
            keys = extract_translation_keys(content)
            used_keys.update(keys)
            
            hardcoded = find_hardcoded_strings(content)
            if hardcoded:
                rel_path = os.path.relpath(file_path, SRC_DIR)
                hardcoded_strings[rel_path] = hardcoded
        except Exception as e:
            print(f"Warning: Error reading {file_path}: {e}")
    
    print(f'Found {len(used_keys)} unique translation keys in code\n')
    
    print('🔍 Checking for missing keys...')
    for used_key in used_keys:
        parsed = parse_translation_key(used_key)
        namespace = parsed['namespace']
        key = parsed['key']
        mapped_path = NS_MAPPING.get(namespace, namespace)
        
        for lang in LANGUAGES:
            locale_data = all_locale_keys[lang].get(namespace)
            
            if not locale_data:
                file_path = LOCALES_DIR / lang / f"{mapped_path}.json"
                if file_path.exists():
                    try:
                        with open(file_path, 'r', encoding='utf-8') as f:
                            locale_data = json.load(f)
                        all_locale_keys[lang][namespace] = locale_data
                    except Exception as e:
                        pass
            
            if not locale_data:
                missing_keys[lang][namespace].append(key)
            else:
                value = get_nested_value(locale_data, key)
                if value is None:
                    missing_keys[lang][namespace].append(key)
    
    print('🔍 Checking for legacy keys...')
    for lang in LANGUAGES:
        for namespace, locale_data in all_locale_keys[lang].items():
            all_keys_in_file = get_all_keys_from_object(locale_data)
            
            for full_key in all_keys_in_file:
                found = False
                for used_key in used_keys:
                    parsed = parse_translation_key(used_key)
                    if parsed['namespace'] == namespace and parsed['key'] == full_key:
                        found = True
                        break
                
                if not found:
                    legacy_keys[lang][namespace].append(full_key)
    
    print('\n' + '=' * 80)
    print('📊 INSPECTION RESULTS')
    print('=' * 80 + '\n')
    
    print('❌ MISSING KEYS:\n')
    has_missing = False
    missing_by_lang = {}
    for lang in sorted(missing_keys.keys()):
        lang_total = 0
        for ns in sorted(missing_keys[lang].keys()):
            if missing_keys[lang][ns]:
                has_missing = True
                lang_total += len(missing_keys[lang][ns])
                if lang not in missing_by_lang:
                    missing_by_lang[lang] = {}
                missing_by_lang[lang][ns] = len(missing_keys[lang][ns])
        if lang_total > 0:
            print(f'  {lang}: {lang_total} missing keys')
    
    if has_missing:
        print('\n  Breakdown by namespace (showing top 10 per language):')
        for lang in sorted(missing_keys.keys()):
            ns_entries = sorted(
                [(ns, len(keys)) for ns, keys in missing_keys[lang].items() if keys],
                key=lambda x: x[1],
                reverse=True
            )[:10]
            if ns_entries:
                print(f'\n    {lang}:')
                for ns, count in ns_entries:
                    print(f'      {ns}: {count} missing keys')
        print('\n  (Full list available in translation-missing-keys.csv)\n')
    else:
        print('  ✅ No missing keys found!\n')
    
    print('🗑️  LEGACY KEYS (not used in code):\n')
    has_legacy = False
    legacy_by_lang = {}
    for lang in sorted(legacy_keys.keys()):
        lang_total = 0
        for ns in sorted(legacy_keys[lang].keys()):
            if legacy_keys[lang][ns]:
                has_legacy = True
                lang_total += len(legacy_keys[lang][ns])
                if lang not in legacy_by_lang:
                    legacy_by_lang[lang] = {}
                legacy_by_lang[lang][ns] = len(legacy_keys[lang][ns])
        if lang_total > 0:
            print(f'  {lang}: {lang_total} legacy keys')
    
    if has_legacy:
        print('\n  Top namespaces with legacy keys:')
        for lang in sorted(legacy_keys.keys()):
            ns_entries = sorted(
                [(ns, len(keys)) for ns, keys in legacy_keys[lang].items() if keys],
                key=lambda x: x[1],
                reverse=True
            )[:5]
            if ns_entries:
                print(f'\n    {lang}:')
                for ns, count in ns_entries:
                    print(f'      {ns}: {count} legacy keys')
        print('\n  (Full list available in translation-inspection-report.json)\n')
    else:
        print('  ✅ No legacy keys found!\n')
    
    print('⚠️  POTENTIALLY HARDCODED STRINGS:\n')
    hardcoded_files = sorted(hardcoded_strings.keys())
    if hardcoded_files:
        total_hardcoded = sum(len(items) for items in hardcoded_strings.values())
        print(f'  Found {total_hardcoded} potential hardcoded strings across {len(hardcoded_files)} files\n')
        print('  Top 20 files with most hardcoded strings:')
        sorted_files = sorted(
            [(file, len(hardcoded_strings[file])) for file in hardcoded_files],
            key=lambda x: x[1],
            reverse=True
        )[:20]
        
        for file, count in sorted_files:
            print(f'    {file}: {count} occurrences')
            for item in hardcoded_strings[file][:3]:
                text_preview = item['text'][:60] + ('...' if len(item['text']) > 60 else '')
                print(f'      Line {item["line"]}: "{text_preview}"')
            if len(hardcoded_strings[file]) > 3:
                print(f'      ... and {len(hardcoded_strings[file]) - 3} more')
            print()
        if len(hardcoded_files) > 20:
            print(f'  ... and {len(hardcoded_files) - 20} more files with hardcoded strings\n')
    else:
        print('  ✅ No hardcoded strings detected!\n')
    
    print('=' * 80)
    print('✅ Inspection complete!')
    print('=' * 80)
    
    missing_count = sum(len(keys) for lang in missing_keys.values() for keys in lang.values())
    legacy_count = sum(len(keys) for lang in legacy_keys.values() for keys in lang.values())
    
    print('\n📈 SUMMARY:')
    print(f'  Total translation keys used: {len(used_keys)}')
    print(f'  Missing keys: {missing_count}')
    print(f'  Legacy keys: {legacy_count}')
    print(f'  Files with hardcoded strings: {len(hardcoded_files)}')
    
    import json
    from datetime import datetime
    
    report_data = {
        'summary': {
            'totalUsedKeys': len(used_keys),
            'missingKeysCount': missing_count,
            'legacyKeysCount': legacy_count,
            'hardcodedFilesCount': len(hardcoded_files)
        },
        'missingKeys': {lang: {ns: keys for ns, keys in data.items()} 
                       for lang, data in missing_keys.items()},
        'legacyKeys': {lang: {ns: keys for ns, keys in data.items()} 
                      for lang, data in legacy_keys.items()},
        'hardcodedStrings': hardcoded_strings,
        'usedKeys': sorted(list(used_keys)),
        'timestamp': datetime.now().isoformat()
    }
    
    report_path = Path(__file__).parent / 'translation-inspection-report.json'
    with open(report_path, 'w', encoding='utf-8') as f:
        json.dump(report_data, f, indent=2, ensure_ascii=False)
    print(f'\n📄 Detailed report saved to: {report_path}')
    
    csv_report_path = Path(__file__).parent / 'translation-missing-keys.csv'
    with open(csv_report_path, 'w', encoding='utf-8', newline='') as f:
        f.write('Language,Namespace,Key\n')
        for lang in sorted(missing_keys.keys()):
            for ns in sorted(missing_keys[lang].keys()):
                for key in missing_keys[lang][ns]:
                    f.write(f'{lang},{ns},{key}\n')
    print(f'📊 Missing keys CSV saved to: {csv_report_path}')

if __name__ == '__main__':
    main()

File: inspect-translations.js
const fs = require('fs');
const path = require('path');

const LOCALES_DIR = path.join(__dirname, 'public', 'locales');
const SRC_DIR = path.join(__dirname, 'src');
const LANGUAGES = ['en', 'fr', 'de', 'it'];
const NS_MAPPING = {
  home: 'pages/home',
  blog: 'pages/blog',
  about: 'pages/about',
  faq: 'pages/faq',
  facilities: 'pages/facilities',
  professionals: 'pages/professionals',
  blogArticles: 'pages/blog',
  contact: 'pages/contact',
  privacy: 'legal/privacy',
  terms: 'legal/terms',
  common: 'common',
  admin: 'dashboard/admin',
  dashboard: 'dashboard/dashboard',
  calendar: 'dashboard/calendar',
  dashboardPersonal: 'dashboard/personalDashboard',
  dashboardProfile: 'dashboard/profile',
  messages: 'dashboard/messages',
  marketplace: 'dashboard/marketplace',
  organization: 'dashboard/organization',
  payroll: 'dashboard/payroll',
  contracts: 'dashboard/contracts',
  dropdowns: 'dropdowns',
  validation: 'validation',
  auth: 'auth',
  'pages/faq': 'pages/faq',
  sitemap: 'pages/sitemap'
};

function getAllFiles(dirPath, arrayOfFiles = []) {
  const files = fs.readdirSync(dirPath);
  files.forEach(file => {
    const filePath = path.join(dirPath, file);
    if (fs.statSync(filePath).isDirectory()) {
      if (!filePath.includes('node_modules') && !filePath.includes('.git')) {
        getAllFiles(filePath, arrayOfFiles);
      }
    } else if (file.endsWith('.js') || file.endsWith('.jsx')) {
      arrayOfFiles.push(filePath);
    }
  });
  return arrayOfFiles;
}

function extractTranslationKeys(content) {
  const keys = new Set();
  
  const patterns = [
    /t\(['"`]([^'"`]+)['"`]/g,
    /t\(['"`]([^'"`]+)['"`]\s*,\s*['"`]/g,
    /i18n\.t\(['"`]([^'"`]+)['"`]/g,
    /i18next\.t\(['"`]([^'"`]+)['"`]/g,
    /\.t\(['"`]([^'"`]+)['"`]/g,
  ];
  
  patterns.forEach(pattern => {
    let match;
    while ((match = pattern.exec(content)) !== null) {
      const key = match[1];
      if (key && 
          !key.includes('{{') && 
          !key.includes('${') &&
          !key.startsWith('http') &&
          key.length > 0) {
        keys.add(key);
      }
    }
  });
  
  return Array.from(keys);
}

function loadLocaleFile(lang, namespace) {
  const filePath = path.join(LOCALES_DIR, lang, `${namespace}.json`);
  if (fs.existsSync(filePath)) {
    try {
      return JSON.parse(fs.readFileSync(filePath, 'utf8'));
    } catch (e) {
      console.warn(`Error loading ${filePath}:`, e.message);
      return null;
    }
  }
  return null;
}

function getAllKeysFromObject(obj, prefix = '') {
  const keys = [];
  for (const key in obj) {
    const fullKey = prefix ? `${prefix}.${key}` : key;
    if (typeof obj[key] === 'object' && obj[key] !== null && !Array.isArray(obj[key])) {
      keys.push(...getAllKeysFromObject(obj[key], fullKey));
    } else {
      keys.push(fullKey);
    }
  }
  return keys;
}

function getNestedValue(obj, keyPath) {
  const keys = keyPath.split('.');
  let value = obj;
  for (const key of keys) {
    if (value && typeof value === 'object' && key in value) {
      value = value[key];
    } else {
      return undefined;
    }
  }
  return value;
}

function parseTranslationKey(key) {
  if (key.includes(':')) {
    const colonIndex = key.indexOf(':');
    const namespace = key.substring(0, colonIndex).trim();
    const keyPath = key.substring(colonIndex + 1).trim();
    return {
      namespace: namespace,
      key: keyPath
    };
  }
  return {
    namespace: 'common',
    key: key
  };
}

function findHardcodedStrings(content) {
  const hardcoded = [];
  const lines = content.split('\n');
  
  const suspiciousPatterns = [
    /(?:title|label|placeholder|text|message|error|success|warning|info|description|button|submit|cancel|save|delete|edit|add|remove|close|open|confirm|reject|approve|decline|accept|logout|login|register|forgot|reset|change|profile|settings|help|support|contact|about|terms|privacy|language|theme|notifications|messages|dashboard|home|back|next|continue|finish|required|optional|select|search|filter|sort|show|hide|view|preview|download|upload|expand|collapse|enable|disable|activate|deactivate)\s*[:=]\s*['"`]([A-Z][^'"`]{3,})['"`]/gi,
    />\s*([A-Z][a-zA-Z\s]{3,})\s*</g,
    /['"`]([A-Z][a-zA-Z\s]{5,})['"`]/g
  ];
  
  lines.forEach((line, index) => {
    if (line.trim().startsWith('//') || line.trim().startsWith('*')) return;
    
    suspiciousPatterns.forEach(pattern => {
      let match;
      while ((match = pattern.exec(line)) !== null) {
        const text = match[1] || match[0];
        if (text && 
            !text.includes('http') && 
            !text.includes('className') && 
            !text.includes('id=') &&
            !text.includes('src=') &&
            !text.match(/^[A-Z][a-z]+[A-Z]/) &&
            text.length > 3 &&
            !text.includes('{{') &&
            !text.includes('${')) {
          hardcoded.push({
            line: index + 1,
            text: text.substring(0, 100),
            context: line.trim().substring(0, 150)
          });
        }
      }
    });
  });
  
  return hardcoded;
}

function main() {
  console.log('🔍 Starting translation inspection...\n');
  
  const allLocaleKeys = {};
  const usedKeys = new Set();
  const missingKeys = {};
  const legacyKeys = {};
  const hardcodedStrings = {};
  
  console.log('📚 Loading locale files...');
  LANGUAGES.forEach(lang => {
    allLocaleKeys[lang] = {};
    Object.entries(NS_MAPPING).forEach(([ns, filePath]) => {
      const data = loadLocaleFile(lang, filePath);
      if (data) {
        allLocaleKeys[lang][ns] = data;
      }
    });
    
    const commonFile = loadLocaleFile(lang, 'common');
    if (commonFile) {
      allLocaleKeys[lang]['common'] = commonFile;
    }
  });
  
  console.log('🔎 Scanning source files for translation keys...');
  const sourceFiles = getAllFiles(SRC_DIR);
  console.log(`Found ${sourceFiles.length} source files\n`);
  
  sourceFiles.forEach(filePath => {
    try {
      const content = fs.readFileSync(filePath, 'utf8');
      const keys = extractTranslationKeys(content);
      keys.forEach(key => usedKeys.add(key));
      
      const hardcoded = findHardcodedStrings(content);
      if (hardcoded.length > 0) {
        const relativePath = path.relative(SRC_DIR, filePath);
        hardcodedStrings[relativePath] = hardcoded;
      }
    } catch (e) {
      console.warn(`Error reading ${filePath}:`, e.message);
    }
  });
  
  console.log(`Found ${usedKeys.size} unique translation keys in code\n`);
  
  console.log('🔍 Checking for missing keys...');
  usedKeys.forEach(usedKey => {
    const { namespace, key } = parseTranslationKey(usedKey);
    const mappedPath = NS_MAPPING[namespace] || namespace;
    
    LANGUAGES.forEach(lang => {
      let localeData = allLocaleKeys[lang][namespace];
      
      if (!localeData) {
        const filePath = path.join(LOCALES_DIR, lang, `${mappedPath}.json`);
        if (fs.existsSync(filePath)) {
          try {
            localeData = JSON.parse(fs.readFileSync(filePath, 'utf8'));
            allLocaleKeys[lang][namespace] = localeData;
          } catch (e) {
            console.warn(`Error loading ${filePath}:`, e.message);
          }
        }
      }
      
      if (!localeData) {
        if (!missingKeys[lang]) missingKeys[lang] = {};
        if (!missingKeys[lang][namespace]) missingKeys[lang][namespace] = [];
        missingKeys[lang][namespace].push(key);
      } else {
        const value = getNestedValue(localeData, key);
        if (value === undefined) {
          if (!missingKeys[lang]) missingKeys[lang] = {};
          if (!missingKeys[lang][namespace]) missingKeys[lang][namespace] = [];
          missingKeys[lang][namespace].push(key);
        }
      }
    });
  });
  
  console.log('🔍 Checking for legacy keys...');
  LANGUAGES.forEach(lang => {
    Object.keys(allLocaleKeys[lang]).forEach(namespace => {
      const localeData = allLocaleKeys[lang][namespace];
      const allKeysInFile = getAllKeysFromObject(localeData);
      
      allKeysInFile.forEach(fullKey => {
        let found = false;
        usedKeys.forEach(usedKey => {
          const { namespace: usedNs, key: usedKeyPath } = parseTranslationKey(usedKey);
          if (usedNs === namespace && usedKeyPath === fullKey) {
            found = true;
          }
        });
        
        if (!found) {
          if (!legacyKeys[lang]) legacyKeys[lang] = {};
          if (!legacyKeys[lang][namespace]) legacyKeys[lang][namespace] = [];
          legacyKeys[lang][namespace].push(fullKey);
        }
      });
    });
  });
  
  console.log('\n' + '='.repeat(80));
  console.log('📊 INSPECTION RESULTS');
  console.log('='.repeat(80) + '\n');
  
  console.log('❌ MISSING KEYS:\n');
  let hasMissing = false;
  const missingByLang = {};
  Object.keys(missingKeys).forEach(lang => {
    let langTotal = 0;
    Object.keys(missingKeys[lang]).forEach(ns => {
      if (missingKeys[lang][ns].length > 0) {
        hasMissing = true;
        langTotal += missingKeys[lang][ns].length;
        if (!missingByLang[lang]) missingByLang[lang] = {};
        missingByLang[lang][ns] = missingKeys[lang][ns].length;
      }
    });
    if (langTotal > 0) {
      console.log(`  ${lang}: ${langTotal} missing keys`);
    }
  });
  
  if (hasMissing) {
    console.log('\n  Breakdown by namespace (showing top 10 per language):');
    Object.keys(missingKeys).sort().forEach(lang => {
      const nsEntries = Object.entries(missingKeys[lang])
        .filter(([ns, keys]) => keys.length > 0)
        .sort((a, b) => b[1].length - a[1].length)
        .slice(0, 10);
      if (nsEntries.length > 0) {
        console.log(`\n    ${lang}:`);
        nsEntries.forEach(([ns, keys]) => {
          console.log(`      ${ns}: ${keys.length} missing keys`);
        });
      }
    });
    console.log('\n  (Full list available in translation-missing-keys.csv)\n');
  } else {
    console.log('  ✅ No missing keys found!\n');
  }
  
  console.log('🗑️  LEGACY KEYS (not used in code):\n');
  let hasLegacy = false;
  const legacyByLang = {};
  Object.keys(legacyKeys).forEach(lang => {
    let langTotal = 0;
    Object.keys(legacyKeys[lang]).forEach(ns => {
      if (legacyKeys[lang][ns].length > 0) {
        hasLegacy = true;
        langTotal += legacyKeys[lang][ns].length;
        if (!legacyByLang[lang]) legacyByLang[lang] = {};
        legacyByLang[lang][ns] = legacyKeys[lang][ns].length;
      }
    });
    if (langTotal > 0) {
      console.log(`  ${lang}: ${langTotal} legacy keys`);
    }
  });
  
  if (hasLegacy) {
    console.log('\n  Top namespaces with legacy keys:');
    Object.keys(legacyKeys).sort().forEach(lang => {
      const nsEntries = Object.entries(legacyKeys[lang])
        .filter(([ns, keys]) => keys.length > 0)
        .sort((a, b) => b[1].length - a[1].length)
        .slice(0, 5);
      if (nsEntries.length > 0) {
        console.log(`\n    ${lang}:`);
        nsEntries.forEach(([ns, keys]) => {
          console.log(`      ${ns}: ${keys.length} legacy keys`);
        });
      }
    });
    console.log('\n  (Full list available in translation-inspection-report.json)\n');
  } else {
    console.log('  ✅ No legacy keys found!\n');
  }
  
  console.log('⚠️  POTENTIALLY HARDCODED STRINGS:\n');
  const hardcodedFiles = Object.keys(hardcodedStrings);
  if (hardcodedFiles.length > 0) {
    const totalHardcoded = Object.values(hardcodedStrings).reduce((sum, arr) => sum + arr.length, 0);
    console.log(`  Found ${totalHardcoded} potential hardcoded strings across ${hardcodedFiles.length} files\n`);
    console.log('  Top 20 files with most hardcoded strings:');
    const sortedFiles = hardcodedFiles
      .map(file => ({ file, count: hardcodedStrings[file].length }))
      .sort((a, b) => b.count - a.count)
      .slice(0, 20);
    
    sortedFiles.forEach(({ file, count }) => {
      console.log(`    ${file}: ${count} occurrences`);
      hardcodedStrings[file].slice(0, 3).forEach(item => {
        console.log(`      Line ${item.line}: "${item.text.substring(0, 60)}${item.text.length > 60 ? '...' : ''}"`);
      });
      if (hardcodedStrings[file].length > 3) {
        console.log(`      ... and ${hardcodedStrings[file].length - 3} more`);
      }
      console.log();
    });
    if (hardcodedFiles.length > 20) {
      console.log(`  ... and ${hardcodedFiles.length - 20} more files with hardcoded strings\n`);
    }
  } else {
    console.log('  ✅ No hardcoded strings detected!\n');
  }
  
  console.log('='.repeat(80));
  console.log('✅ Inspection complete!');
  console.log('='.repeat(80));
  
  const summary = {
    totalUsedKeys: usedKeys.size,
    missingKeysCount: Object.values(missingKeys).reduce((sum, lang) => 
      sum + Object.values(lang).reduce((s, keys) => s + keys.length, 0), 0),
    legacyKeysCount: Object.values(legacyKeys).reduce((sum, lang) => 
      sum + Object.values(lang).reduce((s, keys) => s + keys.length, 0), 0),
    hardcodedFilesCount: hardcodedFiles.length
  };
  
  console.log('\n📈 SUMMARY:');
  console.log(`  Total translation keys used: ${summary.totalUsedKeys}`);
  console.log(`  Missing keys: ${summary.missingKeysCount}`);
  console.log(`  Legacy keys: ${summary.legacyKeysCount}`);
  console.log(`  Files with hardcoded strings: ${summary.hardcodedFilesCount}`);
  
  const reportData = {
    summary,
    missingKeys,
    legacyKeys,
    hardcodedStrings,
    usedKeys: Array.from(usedKeys).sort(),
    timestamp: new Date().toISOString()
  };
  
  const reportPath = path.join(__dirname, 'translation-inspection-report.json');
  fs.writeFileSync(reportPath, JSON.stringify(reportData, null, 2), 'utf8');
  console.log(`\n📄 Detailed report saved to: ${reportPath}`);
  
  const csvReportPath = path.join(__dirname, 'translation-missing-keys.csv');
  let csvContent = 'Language,Namespace,Key\n';
  Object.keys(missingKeys).sort().forEach(lang => {
    Object.keys(missingKeys[lang]).sort().forEach(ns => {
      missingKeys[lang][ns].forEach(key => {
        csvContent += `${lang},${ns},${key}\n`;
      });
    });
  });
  fs.writeFileSync(csvReportPath, csvContent, 'utf8');
  console.log(`📊 Missing keys CSV saved to: ${csvReportPath}`);
}

if (require.main === module) {
  main();
}

module.exports = { main };

